### DISCUSSION TITLE
Docker Swarm and Nginx Load Balancer Confusion

### DISCUSSION THREAD

**Discussion by omairnabiel** [BretFisher/ama#155] [Q&A]
Hi Bret,
Hope you are well. I took your Docker mastery course on Udemy and got to say it's one the best out there. Many few can explain like you do. Incredible. I'm got a little confused when I was introduced to docker swarm load balancing. I cannot seem to understand that if docker swarm can load balance our requests does that take out the need of Nginx load balancing from the picture or they can be used together? How would they work together if they may? Thanks alot

**Comment by BretFisher** [comment]
Hey there,

I like to think of the different types (or layers) of traffic that you deal with in a container cluster:

**1. container-to-container traffic on the same docker network.** In Swarm, this is usually an overlay network, and each service has a VIP that distributes new connections among the tasks of a single service. Swarm provides the DNS service discovery as well, so no Nginx is needed here.

**2. Connections coming into a single Swarm node from external sources.** In this case, you've published a port for a service so that external services/clients can access that Swarm service. Connections are coming into a Swarm node and hitting the routing mesh on the overlay network and will direct the traffic to the VIP of the service that is published on that public port. A few things this *doesn't* do are:
    1. DNS name-based routing on that external port (layer 7 reverse proxy). If you published on port 80 in the Swarm, that's the *only* service in the Swarm that can use that external IP+port. Most Swarms end up with many web sites/API's needing publishing on 80/443.
    2. Node load ballancing for incoming connections to Swarm. The above assumes all connection are coming into the single node we're discussing, but you likely want to distribute incoming connections accross multiple nodes to proivde high availaibilyt and load ballancing for new connectings hitting the routing mesh.

**To solve 1. above,** you would use a "Swarm aware proxy" as an additional service. This would be the service that listens on 80/443 and routes connections to the proper Swarm service over an overlay network using DNS names, or "host header routing".  It add's a new hop for packets of incoming connections, and is an unfortunate side affect of hosting many containers on the same servers that need to listen on the same port. Typically, cloud server instances don't let you have a dozen IP's for a server (so one could be dedicated to each service), so this is our way around that limitation. [Traefik](https://traefik.io/) and [Docker Flow Proxy](https://proxy.dockerflow.com/) are good examples of "Swarm aware proxies".

**You can see me [demoing this from DockerCon 2018](https://dockercon2018.hubs.vidyard.com/watch/k3Cv676wmxAwYDxbvcgcgC) and my [demo repo for doing all this with Traefik is over here](https://github.com/BretFisher/dogvscat).**

**To solve 2. above,** you'd need either an external proxy/LB (AWS ELB, Digital Ocean LB, Azure LB, etc.) or you could potentally use DNS round-robin if the the only external connections you need are browsers (which are smart enough to see multiple DNS A records for a website and properly fail over if one IP is down).

NOTE: Special use cases. In some cases, like PHP and ColdFusion, Nginx acts as the HTTP listener service for those language processors. Those aren't related to above, and if you were someone who used something like "PHP_FPM + Nginx" Then you'd likely have a 1-for-1 relationship of PHP to Nginx, in addition to the traffic layers discussed above, but that's out of the scope of this question :)

**Comment by al-yakubovich** [comment]
@BretFisher 

It is still not clear to me why we cannot use Nginx for load balancing of incoming connections to Swarm.

For example, we can use the following config for Nginx (nginx.conf):

```
upstream loadbalancer {
server 172.17.0.1:5001 weight=6;
server 172.17.0.1:5002 weight=4;
}
server {
location / {
proxy_pass http://loadbalancer;
}}
```

**Comment by BretFisher** [comment]
Where did I say you can't use nginx? Assuming it's running in your Swarm cluster, and listening on 80/443, it can be the host-header traffic router for ingress inter-cluster connections.

**Comment by al-yakubovich** [comment]
Thanks for the answer. Probably I just don't understand load balancers for swarm and have an incorrect interpretation of what I've read.

Let me try again. You said:

"Connections are coming into a Swarm node and hitting the routing mesh on the overlay network and will direct the traffic to the VIP of the service that is published on that public port. A few things this doesn't do are:
....
2. Node load ballancing for incoming connections to Swarm. The above assumes all connection are coming into the single node we're discussing, but you likely want to distribute incoming connections accross multiple nodes to proivde high availaibilyt and load ballancing for new connectings hitting the routing mesh."

Does an ingress network do round-robin across multiple nodes by default? Form statement above it looks like that no (at least for me). But I pretty sure it should do round-robin across multiple nodes (or maybe I don't understand what ingress network is).

Also, I don't understand what is the difference between Nginx load balancer, HAProxy, and Traefik for load balancing. Can I use them interchangeably to do load balancing? Do I need them if I have only REST APIs endpoints?

**Comment by BretFisher** [comment]
does publishing a port in Swarm create a round-robin load balancer? Yes If you publish port 80 in a Swarm service, then that port will be listening on every node, and will round-robin connections (it's technically random but we'll say round-robin for simplicity) between all the containers of that Swarm Service.

Ingress isn't a Swarm feature, so when I say "ingress" I'm referring to the generic definition: incoming connections to a node on your cluster. If that listening port, let's say port 80, is being used by a service running a single NGINX container, then the NGINX proxy will decide what to do with that traffic.

There are many apps that can act as a HTTP proxy. Apache, NGINX, HAProxy, Traefik, and many more. They all have unique features, but all provide the basic function of "receive a HTTP request, and proxy that connection to some other service".

You almost always need a proxy that does HTTP host-header routing. Only one app can listen to port 80 on a single IP address, so if you have a Swarm cluster with more than one public HTTP/S endpoint, you'll need a proxy to listen on 80/443 and decided based on host headers where to send the connection.

That isn't unique to containers, that's just distributed and cloud computing norms.
